lib/cloud/gcp/sql.go

```diff
@@ -99,42 +97,41 @@ func (g *gcpSQLAdminClient) GetDatabaseInstance(ctx context.Context, db types.Da
 	return dbi, nil
 }
 
-// GenerateEphemeralCert returns a new client certificate with RSA key created
-// using the GenerateEphemeralCertRequest Cloud SQL API. Client certificates are
-// required when enabling SSL in Cloud SQL.
-func (g *gcpSQLAdminClient) GenerateEphemeralCert(ctx context.Context, db types.Database, certExpiry time.Time) (*tls.Certificate, error) {
+// GenerateEphemeralCert returns a new client certificate created using the
+// GenerateEphemeralCertRequest Cloud SQL API. Client certificates are required
+// when enabling SSL in Cloud SQL.
+func (g *gcpSQLAdminClient) GenerateEphemeralCert(ctx context.Context, db types.Database, certExpiry time.Time, pubKey crypto.PublicKey) (string, error) {
 	// TODO(jimbishopp): cache database certificates to avoid expensive generate
 	// operation on each connection.
 
-	// Generate RSA private key, x509 encoded public key, and append to certificate request.
-	pkey, err := rsa.GenerateKey(rand.Reader, constants.RSAKeySize)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	pkix, err := x509.MarshalPKIXPublicKey(pkey.Public())
-	if err != nil {
-		return nil, trace.Wrap(err)
+	var keyPEM []byte
+	switch pubKey.(type) {
+	case *rsa.PublicKey:
+		// keys.MarshalPublicKey would use PKCS#1 format for an RSA public key,
+		// we specifically want PKIX here.
+		pkix, err := x509.MarshalPKIXPublicKey(pubKey)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		keyPEM = pem.EncodeToMemory(&pem.Block{Bytes: pkix, Type: "RSA PUBLIC KEY"})
+	default:
+		var err error
+		keyPEM, err = keys.MarshalPublicKey(pubKey)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
```

Comment:
Shouldn't either `x509.MarshalPKIXPublicKey` or `keys.MarshalPublicKey` handle all types?

Reply:
here's what's going on - everything is actually still going to be RSA, and what we have here is working so I don't want to change it - but it's fairly non-standard to have a PEM block of type `RSA PUBLIC KEY` with a PKIX DER-encoded RSA key.

Usually you'll use `RSA PUBLIC KEY` with a PKCS#1 DER encoding, or else just `PUBLIC KEY` with a PKIX DER encoding.

`key.MarshalPublicKey` does 1 of those 2 options I'd expect to be correct. I just have a special case here for RSA keys to keep it unchanged because I'm not actually testing this against GCP